python/hackathon_app/sheets.py

```diff
@@ -21,121 +25,170 @@ def __init__(self, spreadsheet_id: str, cred_file: str):
+        return result
+
+    def register_user(self, hackathon: str, user: "User"):
         """Register user to a hackathon"""
-        if not self._is_created(email):
-            self.create_user(
-                first_name, last_name, email, company, country, tshirt_size
-            )
-        if not self._is_registered(email, hackathon):
-            self._register_user(email, hackathon)
-
-    def _is_created(self, email: str) -> bool:
-        """Check if user already exists in the users sheet."""
-        users = self.get_users()
+        if not self.users.is_created(user):
+            self.users.create(user)
+        registrant = Registrant(
+            user_email=user.email,
+            hackathon_name=hackathon,
+            date_registered=datetime.datetime.now(),
+            attended=None,
+        )
+        if not self.registrations.is_registered(registrant):
+            self.registrations.register(registrant)
+
+
+class WhollySheet:
+    def __init__(self, client, spreadsheet_id, sheet_name):
+        self.client = client
+        self.spreadsheet_id = spreadsheet_id
+        self.range = sheet_name + "!A1:end"
+
+    def insert(self, data):
+        """Insert data as rows into sheet"""
+        try:
+            serialized_ = cattr.unstructure(data)
+            serialized = self._convert_to_list(serialized_)
+            self.client.append(
+                spreadsheetId=self.spreadsheet_id,
+                range=self.range,
+                insertDataOptions="INSERT_ROWS",
+                valueInputOption="RAW",
+                body=serialized,
+            ).execute()
+        except e:
+            print("Oops. No go.")
+
+    def rows(self, structure):  # -> TStructure:
+        """Retrieve rows from sheet"""
+        try:
+            response = self.client.get(
+                spreadsheetId=self.spreadsheet_id, range=self.range
+            ).execute()
+            rows = response["values"]
+            data = self._convert_to_dict(rows)
+            response = cattr.structure(data, structure)
+        except (TypeError, AttributeError):
+            raise DeserializeError("Bad Data")
+        return response
+
+    def _convert_to_dict(self, data) -> Sequence[Dict[str, str]]:
+        """Given a list of lists where the first list contains key names, convert it to
+        a list of dictionaries.
+        """
+        result: Sequence[Dict[str, str]] = [dict(zip(data[0], r)) for r in data[1:]]
+        return result
+
+    def _convert_to_list(
+        self, data: Dict[str, Union[str, int, Sequence[str], datetime.datetime, None]]
+    ) -> Sequence:
+        """Given a dictionary, return a list containing its values"""
+        return list(data.values())
+
+
+@attr.s(auto_attribs=True)
+class User:
+    first_name: str
+    last_name: str
+    email: str
+    date_created: Optional[datetime.datetime]
+    organization: str
+    tshirt_size: str
+    dietary_restrictions: Optional[List[str]]
```

Comment:
I know @jkaster wants this but I'm wondering if we can punt till at least we have a working MVP up and deployed? I think the proper way to handle this field would be a multi-select on the front-end  => comma separated string in a single cell on the google sheet. That won't be too much work but you'll have to special case serialization here.

Here's my suggestion but again I'd prefer to just merge this code w/out the field and add it later after we have the `looker_sdk` calls implemented. Something like:

```python
class User:
    first_name: str
    .
    .
    .
    _dietary_restrictions_data: str = ""  # might not work with leading _ but we'll treat it as "private" anyway

    @property
    def dietary_restrictions(self) -> List[str]:  # is that the correct syntax for typing an @property?
        return self._dietary_restrictions_data.split(",")

    @dietary_restrictions.setter
    def dietary_restrictions(self, restrictions: List[str]) -> None:
        self._dietary_restrictions_data = restrictions.join(",")
```

Reply:
Dietary restrictions is *definitely* not a requirement for the MVP. If it's a hassle, defer it.